libraries/standard/json/utest/json_utest.c

```diff
@@ -0,0 +1,1153 @@
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\u\xD8\x3D\\u\xde\x07\"}}"
+#define ILLEGAL_UNICODE_LITERAL_HEX_LENGTH                  ( sizeof( ILLEGAL_UNICODE_LITERAL_HEX ) - 1 )
+
+#define UNICODE_PREMATURE_LOW_SURROGATE   \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\ude07\\uD83D\"}}"
+#define UNICODE_PREMATURE_LOW_SURROGATE_LENGTH              ( sizeof( UNICODE_PREMATURE_LOW_SURROGATE ) - 1 )
+
+#define UNICODE_VALID_HIGH_NO_LOW_SURROGATE \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY   \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\uD83D. Hello there!\"}}"
+#define UNICODE_VALID_HIGH_NO_LOW_SURROGATE_LENGTH          ( sizeof( UNICODE_VALID_HIGH_NO_LOW_SURROGATE ) - 1 )
+
+#define UNICODE_WRONG_ESCAPE_AFTER_HIGH_SURROGATE \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY         \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\uD83D\\Ude07\"}}"
+#define UNICODE_WRONG_ESCAPE_AFTER_HIGH_SURROGATE_LENGTH    ( sizeof( UNICODE_WRONG_ESCAPE_AFTER_HIGH_SURROGATE ) - 1 )
+
+#define UNICODE_VALID_HIGH_INVALID_LOW_SURROGATE \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY        \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\uD83D\\uEFFF\"}}"
+#define UNICODE_VALID_HIGH_INVALID_LOW_SURROGATE_LENGTH     ( sizeof( UNICODE_VALID_HIGH_INVALID_LOW_SURROGATE ) - 1 )
+
+#define UNICODE_BOTH_SURROGATES_HIGH      \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\uD83D\\uD83D\"}}"
+#define UNICODE_BOTH_SURROGATES_HIGH_LENGTH                 ( sizeof( UNICODE_BOTH_SURROGATES_HIGH ) - 1 )
+
+/* For security, \u0000 is disallowed. */
+#define UNICODE_ESCAPE_SEQUENCE_ZERO_CP   \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\u0000\"}}"
+#define UNICODE_ESCAPE_SEQUENCE_ZERO_CP_LENGTH    ( sizeof( UNICODE_ESCAPE_SEQUENCE_ZERO_CP ) - 1 )
+
+/* /NUL escape is disallowed. */
+#define NUL_ESCAPE                        \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\\x0\"}}"
+#define NUL_ESCAPE_LENGTH           ( sizeof( NUL_ESCAPE ) - 1 )
+
+#define ESCAPE_CHAR_ALONE                 \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\\"}}"
+#define ESCAPE_CHAR_ALONE_LENGTH    ( sizeof( ESCAPE_CHAR_ALONE ) - 1 )
+
+/* Any escape character less than 0x21 is considered illegal. */
+#define UNKNOWN_ESCAPE                    \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\\\x20\"}}"
+#define UNKNOWN_ESCAPE_LENGTH    ( sizeof( UNKNOWN_ESCAPE ) - 1 )
+
+/* An unescaped control character is considered ILLEGAL. */
+#define UNESCAPED_CONTROL_CHAR            \
+    "{\"foo\":\"abc\",\"" FIRST_QUERY_KEY \
+    "\":{\"" SECOND_QUERY_KEY "\" : \"\x15\"}}"
+#define UNESCAPED_CONTROL_CHAR_LENGTH    ( sizeof( UNESCAPED_CONTROL_CHAR ) - 1 )
+
+/* Triggers the case in which i >= max for skipUTF8MultiByte.
```

Comment:
What is 'i' in i>=max?

Reply:
The iterator to advance for each skip function (Added a comment stating this).